lib/msf/core/exploit/browser_autopwnv2.rb

```diff
@@ -0,0 +1,834 @@
+        platform_obj = Msf::Module::Platform.find_platform(payload_platform.to_s)
+      rescue ArgumentError
+        false
+      end
+
+      return true if platform_obj && m.platform.platforms.include?(platform_obj)
+
+      false
+    end
+
+
+    # Checks whether the payload is compatible with the module based on the module's compatibility list
+    #
+    # @param compatible_payloads [Array] A list of payloads that are compatible
+    # @param payload_name [String]
+    # @return [TrueClass] Payload is compatible.
+    # @return [FalseClass] Payload is not compatible.
+    def is_payload_compatible?(compatible_payloads, payload_name)
+      compatible_payloads.each do |k|
+        return true if k[0] == payload_name
+      end
+
+      false
+    end
+
+
+    # Checks if the module is multi-platform based on the directory path.
+    #
+    # @param m [Object] Module.
+    # @return Module [TrueClass] is multi-platform.
+    # @return Module [FalseClass] is not multi-platform.
+    def is_multi_platform_exploit?(m)
+      m.fullname.include?('multi/')
+    end
+
+
+    # Returns an appropriate payload that's compatible with the module.
+    #
+    # @param m [Object] A module that's been initialized.
+    # @return [Array] Payload name. Example: 'windows/meterpreter/reverse_tcp'
+    def select_payload(m)
+      compatible_payloads = []
+
+      module_payloads = nil
+
+      DEFAULT_PAYLOADS.each_pair do |platform, info|
+        payload_choice = {
+          :payload_name => get_selected_payload_name(platform),
+          :payload_lport => get_selected_payload_lport(platform)
+        }
+
+        if !is_multi_platform_exploit?(m) && !m.platform.platforms.empty? && is_payload_platform_compatible?(m, platform)
+          compatible_payloads << payload_choice
+          break
+        else
+          # The #compatible_payloads method is super expensive (slow). By doing it this way,
+          # I managed to shave off seconds.
+          module_payloads ||= m.compatible_payloads
+
+          if is_payload_compatible?(module_payloads, payload_choice[:payload_name])
+            compatible_payloads << payload_choice
+          end
+        end
+      end
+
+      @wanted_payloads.concat(compatible_payloads)
+
+      compatible_payloads
+    end
+
+
+    # Starts exploits.
+    #
+    # @return [void]
+    def start_exploits
+      bap_exploits.each do |m|
+        set_exploit_options(m)
+        m.exploit_simple(
+          'LocalInput'  => self.user_input,
+          'LocalOutput' => self.user_output,
+          'Target'      => 0,
+          'Payload'     => m.datastore['PAYLOAD'],
+          'RunAsJob'    => true
+        )
+        @exploit_job_ids << m.job_id
+      end
+    end
+
+
+    # Sets up BAPv2. This is like our main function.
+    #
+    # @return [void]
+    def setup
+      t1 = Time.now
+      self.datastore['MODULEOWNER'] = Msf::Exploit::Remote::BrowserAutopwnv2
+      self.datastore['DisablePayloadHandler'] = true
```

Comment:
The implementation of this module as an exploit may have other side effects besides the need for the hack above.